crates/dojo-core/src/database/index.cairo

```diff
@@ -61,6 +164,90 @@ fn get(address_domain: u32, index: felt252) -> Array<felt252> {
     res
 }
 
+/// Gets all ids for a given index, as well as all keys for each id.
+/// # Arguments
+/// * address_domain - The address domain to write to.
+/// * index - The index to write to.
+/// # Returns
+/// * ids - The ids for the index.
+fn get_with_keys(address_domain: u32, index: felt252, keys_layout: Span<u8>) -> (Array<felt252>, Array<Span<felt252>>) {
+    let mut ids = ArrayTrait::new();
+    let mut all_keys = ArrayTrait::new();
+
+    let index_len_key = build_index_len_key(index);
+    let index_len = storage::get(address_domain, index_len_key);
+    let mut idx = 0;
+
+    loop {
+        if idx == index_len {
+            break ();
+        }
+
+        let id = storage::get(address_domain, build_index_key(index, idx));
+        let keys_key = build_index_item_keys(index, id);
+
+        let len_felt = (*storage::get_many(address_domain, keys_key, 0, 1, array![8].span()).at(0));
+        let len = len_felt.try_into().unwrap();
+        let mut end_keys: u64 = 8 + 251 * len.into();
+
+        let mut layout = array![8];
+        let mut layout_idx = 0;
+        loop {
+            if layout_idx == len * 2 {
+                break ();
+            }
+            else if layout_idx < len {
+                layout.append(251);
+            } else {
+                layout.append(*keys_layout.at(layout_idx - len));
+                end_keys += (*keys_layout.at(layout_idx - len)).into();
+            }
+            layout_idx += 1;
+        };
+
+        let end: u32 = ((end_keys + 250) / 251).try_into().unwrap();
+
+        let len_pos_and_keys = storage::get_many(address_domain, keys_key, 0, end, layout.span());
+
+        let len = *storage::get_many(address_domain, keys_key, 0, 1, array![8].span()).at(0);
+        let offset: u8 = (len + 1).try_into().unwrap();
+        let len: u32 = (len + len + 1).try_into().unwrap();
+        let keys = storage::get_many(address_domain, keys_key, offset, len, keys_layout);
+        ids.append(id);
+        all_keys.append(keys);
+        idx += 1;
+    };
+
+    (ids, all_keys)
+}
+
+/// Returns all the entries that hold a giben key
+/// # Arguments
+/// * address_domain - The address domain to write to.
+/// * index - The index to read from.
+/// * key - The key return values from.
+fn get_by_key(address_domain: u32, index: felt252, key: felt252) -> Array<felt252> {
+    let mut res = ArrayTrait::new();
+    let specific_len_key = build_index_specific_key_len(index, key);
+    let specific_len = storage::get(address_domain, specific_len_key);
+    let index_len = storage::get(address_domain, specific_len_key);
```

Comment:
`specific_len` and `index_len` are the same?

Reply:
my bad, merged you fix